### PR TITLE
Improve merging of docs on generation

### LINKED ISSUE #9125: Potentially slow code in DocTools::merge_from
### Describe the project you are working on

Godot.

### Describe the problem or limitation you are having in your project

Noticed some potentially slow code in `DocTools::merge_from` https://github.com/godotengine/godot/blob/5f05e2b9b1a3fedcdd7ecb2ab976a2687fd6f19a/editor/doc_tools.cpp#L79. The current algorithm used to merge is `O(n^2)` when it could be `O(n)`, where `n` is the number of members in a class. For example,
for each property in `this`, the following code block scans through all the properties in `p_data` to look for a match.
```cpp
                for (int i = 0; i < c.properties.size(); i++) {
                        DocData::PropertyDoc &p = c.properties.write[i];


                        for (int j = 0; j < cf.properties.size(); j++) {
                                if (cf.properties[j].name != p.name) {
                                        continue;
                                }
                                const DocData::PropertyDoc &pf = cf.properties[j];


                                p.description = pf.description;
                                p.is_deprecated = pf.is_deprecated;
                                p.deprecated_message = pf.deprecated_message;
                                p.is_experimental = pf.is_experimental;
                                p.experimental_message = pf.experimental_message;
                                p.keywords = pf.keywords;
                                break;
                        }
                }
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I haven't profiled to see whether the current algorithm is actually a problem, but I just thought I'd write up a formal issue to bring it to people's attention since the CI runs the doctool on every PR, so making it even a bit more efficient might help a lot. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

First step is profile to check whether it's actually a bottleneck. If it is, I have a few ideas to try out:

1. Start the inner loop index `j` at the current value of the outer loop index `i` on the assumption that `this` and `p_data` usually have the same number of elements in the same order. Optionally, ensure both `DocData` instances are sorted before merging.
2. Store the class members in a `HashSet` or `HashMap` so that the common members can be found more efficiently. Even though this would make the algorithm `O(n)`, I doubt it would actually be faster because of the extra hashing overhead, but it's perhaps worth trying anyway.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

Doctool is core.

### PR DESCRIPTION
**Pull request by AThousandShips**
Cleans up and improves (slightly, it seems) performance in merging docs by using sorted lookup

---

* Closes: https://github.com/godotengine/godot-proposals/issues/9125

### CODE REVIEW

File: core/doc_data.h [hunk 1]
```diff
@@ -122,28 +122,36 @@ class DocData {
...
-					if (arguments.size() == 0 || p_method.arguments.size() == 0) { // 1.
+					// - 3. Constructors starting with class: Foo(Foo, ...)
+					// - 4. Other constructors by arguments
+					if (arguments.is_empty() || p_method.arguments.is_empty()) { // 1.
 						return arguments.size() < p_method.arguments.size();
 					}
-					if (arguments[0].type == return_type || p_method.arguments[0].type == p_method.return_type) { // 2.
-						return (arguments[0].type == return_type) || (p_method.arguments[0].type != p_method.return_type);
+					if (arguments[0].type == return_type || p_method.arguments[0].type == return_type) { // 2 & 3 vs 4.
+						return p_method.arguments[0].type != p_method.return_type;
```
Comment (AThousandShips) [inline core/doc_data.h:145]: The claimed sort above wasn't being done clearly, this enforces that (see below with `Array` and `Color` not following this order)

File: core/doc_data.h [hunk 2]
```diff
@@ -122,28 +122,36 @@ class DocData {
...
 					}
-					return arguments[0] < p_method.arguments[0];
+					for (int i = 0; i < MIN(arguments.size(), p_method.arguments.size()); ++i) {
+						if (arguments[i] < p_method.arguments[i]) {
+							return true;
+						} else if (p_method.arguments[i] < arguments[i]) {
+							return false;
+						}
+					}
+					return arguments.size() < p_method.arguments.size();
```
Comment (AThousandShips) [inline core/doc_data.h:154]: Sorts by arguments more coherently, doesn't change anything currently, but ensures stability going forward and filtering

**Comment by AThousandShips** [comment]
Ran some benchmarking with the following:
```diff
diff --git a/main/main.cpp b/main/main.cpp
index 4c9c159f47..f87e02eb44 100644
--- a/main/main.cpp
+++ b/main/main.cpp
@@ -3257,10 +3257,18 @@ bool Main::start() {
                GLOBAL_DEF(PropertyInfo(Variant::INT, "dotnet/project/assembly_reload_attempts", PROPERTY_HINT_RANGE, "1,16,1,or_greater"), 3);
 #endif

+               OS::get_singleton()->benchmark_begin_measure("Documentation", "Total");
+
+               OS::get_singleton()->benchmark_begin_measure("Documentation", "Generation");
+
                Error err;
                DocTools doc;
                doc.generate(gen_flags);

+               OS::get_singleton()->benchmark_end_measure("Documentation", "Generation");
+
+               OS::get_singleton()->benchmark_begin_measure("Documentation", "Loading");
+
                DocTools docsrc;
                HashMap<String, String> doc_data_classes;
                HashSet<String> checked_paths;
@@ -3299,24 +3307,44 @@ bool Main::start() {
                ERR_FAIL_COND_V_MSG(err != OK, false, "Error loading classes from: " + index_path + ": " + itos(err));
                checked_paths.insert(index_path);

+               OS::get_singleton()->benchmark_end_measure("Documentation", "Loading");
+
+               OS::get_singleton()->benchmark_begin_measure("Documentation", "Merging");
+
                print_line("Merging docs...");
                doc.merge_from(docsrc);

+               OS::get_singleton()->benchmark_end_measure("Documentation", "Merging");
+
+               OS::get_singleton()->benchmark_begin_measure("Documentation", "Erasing");
+
                for (const String &E : checked_paths) {
                        print_line("Erasing old docs at: " + E);
                        err = DocTools::erase_classes(E);
                        ERR_FAIL_COND_V_MSG(err != OK, false, "Error erasing old docs at: " + E + ": " + itos(err));
                }

+               OS::get_singleton()->benchmark_end_measure("Documentation", "Erasing");
+
+               OS::get_singleton()->benchmark_begin_measure("Documentation", "Saving");
+
                print_line("Generating new docs...");
                err = doc.save_classes(index_path, doc_data_classes);
                ERR_FAIL_COND_V_MSG(err != OK, false, "Error saving new docs:" + itos(err));

+               OS::get_singleton()->benchmark_end_measure("Documentation", "Saving");
+
+               OS::get_singleton()->benchmark_begin_measure("Documentation", "Deleting");
+
                print_line("Deleting docs cache...");
                if (FileAccess::exists(EditorHelp::get_cache_full_path())) {
                        DirAccess::remove_file_or_error(EditorHelp::get_cache_full_path());
                }

+               OS::get_singleton()->benchmark_end_measure("Documentation", "Deleting");
+
+               OS::get_singleton()->benchmark_end_measure("Documentation", "Total");
+
                OS::get_singleton()->set_exit_code(EXIT_SUCCESS);
                return false;
        }
```

And for me it gave the following two cases:
```
[Documentation] (Without these changes)
- Generation: 635.507 msec.
- Loading: 382.480 msec.
- Merging: 27.354 msec.
- Erasing: 137.812 msec.
- Saving: 630.196 msec.
- Deleting: 0.716 msec.
- Total: 1814.084 msec.

[Documentation] (With these changes)
- Generation: 635.366 msec.
- Loading: 377.898 msec.
- Merging: 8.881 msec.
- Erasing: 139.279 msec.
- Saving: 634.273 msec.
- Deleting: 0.817 msec.
- Total: 1796.532 msec.
```

Which is marginal but is present, but the generation isn't slowed down by the additional sorting, so I'd say this is an improvement in any case just cleaner code

**Comment by nlupugla** [comment]
Great work! You love to see PRs with more lines removed than added that also speed performance :) I'll go through and review your changes in more detail when I get the chance.

File: doc/classes/Array.xml [hunk 3]
```diff
@@ -52,19 +52,19 @@
...
 		<constructor name="Array">
 			<return type="Array" />
-			<param index="0" name="base" type="Array" />
-			<param index="1" name="type" type="int" />
-			<param index="2" name="class_name" type="StringName" />
-			<param index="3" name="script" type="Variant" />
+			<param index="0" name="from" type="Array" />
 			<description>
-				Creates a typed array from the [param base] array.
+				Returns the same array as [param from]. If you need a copy of the array, use [method duplicate].
```
Comment (Mickeon) [inline doc/classes/Array.xml:57]: Me having to rebase the Array documentation PR for the one billionth time but it's for the best.

File: editor/doc_tools.cpp [hunk 4]
```diff
@@ -96,221 +124,96 @@ void DocTools::merge_from(const DocTools &p_data) {
...
-				m.is_deprecated = mf.is_deprecated;
-				m.deprecated_message = mf.deprecated_message;
-				m.is_experimental = mf.is_experimental;
-				m.experimental_message = mf.experimental_message;
-				m.keywords = mf.keywords;
-				break;
-			}
-		}
+		// Perform ordered merge on names.
+		{
```
Comment (Mickeon) [inline editor/doc_tools.cpp:164]: I think it would be great to do the same thing as `merge_methods` in this PR, and group the other categories as well into their own method. Being left out like this in favour of code regions makes the code look very unwieldy and disorganized. Especially more with short and repeating variable names.

**Comment by nlupugla** [comment]
Since the merging logic is non-trivial now, I might suggesting making a helper function or iterator that you can reuse for the different merge cases. For example, a `merge_vectors` method that takes a function pointer, which you might call like `merge_vectors(to_properties, from_properties, &merge_property)`, where `merge_property` is a function that just handles merging two matched properties.

If you feel that overcomplicates things, I think repeating the iteration code is fine as well.

**Comment by Mickeon** [comment]
I'd say first group them into their own functions (as per above comment). Then, the unification would come natural at a later point, as it would be easier to envision and refactor with "all of the cards on the table".

File: editor/doc_tools.cpp (hunk 4 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:164]: I can put them into their own single methods if desired

**Comment by AThousandShips** [comment]
The comparison isn't the same in the different cases, so it'd be more complicated

Also realized I need to fix some details of this, one moment

**Comment by AThousandShips** [comment]
Fixed some of the code, getting late so will continue working tomorrow, should work now, but with some new changes, will elaborate on them tomorrow

Updated the benchmarks above

**Comment by nlupugla** [comment]
Wow, the new benchmarks show a bigger improvement and it seems like most of it is due to the `saving` somehow getting faster. It doesn't look like any of your code directly affects saving, unless somehow ensuring the docs are sorted makes saving faster. Maybe you your disk drive was less busy during the later tests or something?

**Comment by AThousandShips** [comment]
There's going to be fluctuations especially with file access, will do some more rigorous benchmarking tomorrow, and some more tweaks to the ordering of constructors, it needs a little more fixing but it's almost ready again

File: doc/classes/Array.xml (hunk 3 above)
Comment (AThousandShips) [reply doc/classes/Array.xml]: Current rework removes this change, will see where we land on order, will see :)

**Comment by AThousandShips** [comment]
So preliminary update (will work later today), will drop the operators and constructors for now, they're not difficult to fix as such but they require some discussion on how to handle the ordering etc. that I think it best left to a follow up PR, and they make up a vanishingly small portion of the documentation data (only built-in types have them, and only a few each)

Will push later today and do new benchmarking

**Comment by AThousandShips** [comment]
Some different changes, will elaborate more later, but here are some basic benchmarking for comparison, the difference isn't major, but it's something, and I think the general code improvements are good in themselves as well

```
[Unchanged]
- Generation:   663.865 msec.   660.288 msec.   642.372 msec.      Avrg:   655.508 msec.
- Loading:      414.215 msec.   417.114 msec.   396.416 msec.      Avrg:   409.248 msec.
- Merging:       28.389 msec.    28.027 msec.    28.552 msec.      Avrg:    28.323 msec.
- Erasing:      147.778 msec.   144.207 msec.   136.605 msec.      Avrg:   142.863 msec.
- Saving:       678.956 msec.   699.290 msec.   681.661 msec.      Avrg:   686.636 msec.
- Deleting:       1.001 msec.     1.016 msec.     0.738 msec.      Avrg:     0.918 msec.
- Total:       1934.221 msec.  1949.956 msec.  1886.360 msec.      Avrg:  1923.512 msec.

[With initial]
- Generation:   632.953 msec.   647.692 msec.   645.188 msec.      Avrg:   641.944 msec.
- Loading:      394.408 msec.   407.362 msec.   401.538 msec.      Avrg:   401.103 msec.
- Merging:       28.209 msec.    27.638 msec.    27.739 msec.      Avrg:    27.862 msec.
- Erasing:      122.336 msec.   142.152 msec.   135.156 msec.      Avrg:   133.215 msec.
- Saving:       682.808 msec.   702.175 msec.   707.425 msec.      Avrg:   697.469 msec.
- Deleting:       0.762 msec.     0.848 msec.     0.775 msec.      Avrg:     0.795 msec.
- Total:       1861.495 msec.  1927.882 msec.  1917.837 msec.      Avrg:  1902.405 msec.

[With constructor/operator cleanup]
- Generation:   618.375 msec.   639.833 msec.   645.607 msec.      Avrg:   634.605 msec.
- Loading:      385.870 msec.   388.729 msec.   397.186 msec.      Avrg:   390.595 msec.
- Merging:       15.727 msec.    15.715 msec.    15.666 msec.      Avrg:    15.703 msec.
- Erasing:      141.252 msec.   134.104 msec.   138.175 msec.      Avrg:   137.844 msec.
- Saving:       705.065 msec.   690.901 msec.   683.012 msec.      Avrg:   692.993 msec.
- Deleting:       0.757 msec.     0.723 msec.     1.083 msec.      Avrg:     0.854 msec.
- Total:       1867.062 msec.  1870.022 msec.  1880.743 msec.      Avrg:  1872.609 msec.

[Without Constants]
- Generation:   647.417 msec.   652.177 msec.   648.856 msec.      Avrg:   649.483 msec.
- Loading:      404.471 msec.   397.803 msec.   398.155 msec.      Avrg:   400.143 msec.
- Merging:       16.087 msec.    16.225 msec.    16.278 msec.      Avrg:    16.197 msec.
- Erasing:      143.796 msec.   148.085 msec.   140.245 msec.      Avrg:   144.042 msec.
- Saving:       686.624 msec.   692.332 msec.   708.543 msec.      Avrg:   695.833 msec.
- Deleting:       1.060 msec.     1.100 msec.     1.157 msec.      Avrg:     1.106 msec.
- Total:       1899.472 msec.  1907.738 msec.  1913.250 msec.      Avrg:  1906.820 msec.

[All changes]
- Generation:   626.262 msec.   649.933 msec.   654.115 msec.      Avrg:   643.436 msec.
- Loading:      393.736 msec.   406.013 msec.   402.916 msec.      Avrg:   400.888 msec.
- Merging:       13.201 msec.    12.842 msec.    13.006 msec.      Avrg:    13.016 msec.
- Erasing:      140.544 msec.   143.606 msec.   146.294 msec.      Avrg:   143.481 msec.
- Saving:       728.345 msec.   693.062 msec.   693.980 msec.      Avrg:   705.128 msec.
- Deleting:       0.759 msec.     0.876 msec.     1.025 msec.      Avrg:     0.887 msec.
- Total:       1902.863 msec.  1906.348 msec.  1911.352 msec.      Avrg:  1906.845 msec.
```

**Comment by AThousandShips** [comment]
So some observations and details:
* Constructors are a bit weird (and possibly broken, for a future PR to work out), they only require that all the types match, *in any order* meaning that you can't have `Color(float, Color)` *and* `Color(Color, float)` as separate constructors with separate descriptions
* Operators and constructors are vanishingly few, so no need to process them (at least not for now)
* The results for merging or searching (the current solution) seem to be negligible in testing, so keeping this method, sorting the loaded values should be very cheap, as they are essentially sorted in most cases, but the savings on using a merge is likely still not significant enough compared with sorting, so both methods are essentially equivalent

In the end the improvements for performance seem negligible, though it might be more on some hardware, hard to say, but I think the code style improvements are significant here so IMO this change isn't without merit

Some changes I did:
* Cleaned up the constructor and operator code, it was a bit inefficient, some of the changes:
  - Checking argument count *before* checking names match (kept the check for names in constructors, it's essentially pointless but someone might just add an incorrectly named constructor in the doc file and break things, all the generated constructors will *always* have the same name)
  - Removed the checks in operators for argument matching using the permutation method, as operators (currently) have at most one argument
  - Avoiding mutation checks and locking for the various vectors by loading the data beforehand
  - Fixed index types after upgrade to 64 bits (not going to matter, unless you have impossibly big data, but still good)
* Made some other data improvements and changed the check for plain methods to just check for name order, to avoid accidentally invoking some other ordering

Ready for review now, might need some minor tinkering, but works well now, tested messing with documentation files and it doesn't break when you do things like reordering or deleting entries, the checks are all sufficient

File: editor/doc_tools.cpp
Comment (AThousandShips) [inline editor/doc_tools.cpp:1]: Put these changes in a separate commit to make them easier to review, as I first moved the code out of the main method

File: editor/doc_tools.cpp [hunk 5]
```diff
@@ -76,241 +76,247 @@ static String _translate_doc_string(const String &p_text) {
...
-						if (!arg_used[l]) { // at least one of the arguments was different
-							not_the_same = true;
+			{
+				// Since constructors can repeat, we need to check the type of
+				// the arguments so we make sure they are different.
+				int64_t arg_count = from.arguments.size();
+				Vector<bool> arg_used;
+				arg_used.resize_zeroed(arg_count);
+				// Also there is no guarantee that argument ordering will match, so we
+				// have to check one by one so we make sure we have an exact match.
```
Comment (Mickeon) [inline editor/doc_tools.cpp:112]: Personal peeve
```suggestion
				// Also there is no guarantee that argument ordering will match,
				// so we have to check one by one so we make sure we have an exact match.
```

File: editor/doc_tools.cpp [hunk 6]
```diff
@@ -76,241 +76,247 @@ static String _translate_doc_string(const String &p_text) {
...
-				m.is_deprecated = mf.is_deprecated;
-				m.deprecated_message = mf.deprecated_message;
-				m.is_experimental = mf.is_experimental;
-				m.experimental_message = mf.experimental_message;
-				m.keywords = mf.keywords;
-				break;
-			}
+	SearchArray<DocData::ConstantDoc> search_array;
+
+	for (DocData::ConstantDoc &to : p_to) {
```
Comment (Mickeon) [inline editor/doc_tools.cpp:178]: Can this be `const`?

File: editor/doc_tools.cpp [hunk 7]
```diff
@@ -76,241 +76,247 @@ static String _translate_doc_string(const String &p_text) {
...
-
-				m.description = mf.description;
-				m.is_deprecated = mf.is_deprecated;
-				m.deprecated_message = mf.deprecated_message;
-				m.is_experimental = mf.is_experimental;
-				m.experimental_message = mf.experimental_message;
-				m.keywords = mf.keywords;
-				break;
-			}
+	SearchArray<DocData::ConstantDoc> search_array;
```
Comment (Mickeon) [inline editor/doc_tools.cpp:176]: Can this also be `const`?

File: editor/doc_tools.cpp [hunk 8]
```diff
@@ -76,241 +76,247 @@ static String _translate_doc_string(const String &p_text) {
...
+			to.is_experimental = from.is_experimental;
+			to.experimental_message = from.experimental_message;
+			break;
+		}
+	}
+}
+
+void DocTools::merge_from(const DocTools &p_data) {
+	for (KeyValue<String, DocData::ClassDoc> &E : class_list) {
+		DocData::ClassDoc &c = E.value;
```
Comment (Mickeon) [inline editor/doc_tools.cpp:286]: Unless I am missing something glaring this can be `const` as well?

File: editor/doc_tools.cpp [hunk 9]
```diff
@@ -76,241 +76,247 @@ static String _translate_doc_string(const String &p_text) {
...
+	for (const DocData::ThemeItemDoc &from : p_from) {
+		int64_t found = search_array.bisect(to_ptrw, to_size, from, true);
+
+		if (found >= to_size) {
+			continue;
 		}
 
-		for (int i = 0; i < c.operators.size(); i++) {
-			DocData::MethodDoc &m = c.operators.write[i];
+		DocData::ThemeItemDoc &to = to_ptrw[found];
```
Comment (Mickeon) [inline editor/doc_tools.cpp:241]: And finally, maybe `const`?

File: editor/doc_tools.cpp (hunk 7 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:176]: Doesn't matter, it's a helper class, doesn't change anything

File: editor/doc_tools.cpp (hunk 8 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:286]: Again, written to, see below

File: editor/doc_tools.cpp (hunk 9 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:241]: Ditto

**Review by Mickeon — APPROVED** [review]
Nothing really against this. Even if the performance gain is negligible, what came out of it is greater code quality.

File: editor/doc_tools.cpp [hunk 10]
```diff
@@ -76,241 +76,275 @@ static String _translate_doc_string(const String &p_text) {
...
+		int64_t found = search_array.bisect(to_ptrw, to_size, from, true);
+
+		if (found >= to_size) {
+			continue;
+		}
+
+		DocData::MethodDoc &to = to_ptrw[found];
+
+		// Check found entry on name and argument.
+		if (to.name == from.name && to.arguments.size() == from.arguments.size() && (to.arguments.size() == 0 || to.arguments[0].type == from.arguments[0].type)) {
```
Comment (AThousandShips) [inline editor/doc_tools.cpp:302]:
```suggestion
		if (to.name == from.name && to.arguments.size() == from.arguments.size() && (to.arguments.is_empty() || to.arguments[0].type == from.arguments[0].type)) {
```
Will add later

File: editor/doc_tools.cpp [hunk 11]
```diff
@@ -323,6 +357,8 @@ void DocTools::merge_from(const DocTools &p_data) {
 			for (int j = 0; j < cf.properties.size(); j++) {
 				if (cf.properties[j].name == "GodotSharp") {
 					c.properties.push_back(cf.properties[j]);
+					c.properties.sort();
```
Comment (AThousandShips) [inline editor/doc_tools.cpp:360]: Needed after removing sorting on saving, could use an insert as well, will see later

File: editor/doc_tools.cpp [hunk 12]
```diff
@@ -1696,8 +1744,6 @@ Error DocTools::save_classes(const String &p_default_path, const HashMap<String,
 		_write_method_doc(f, "annotation", c.annotations);
 
 		if (!c.theme_properties.is_empty()) {
-			c.theme_properties.sort();
```
Comment (AThousandShips) [inline editor/doc_tools.cpp:1699]: Removed the sorts here as they were mostly unnecessary and done after already sorting, and moved that sorting into the generation where it does not already apply

File: editor/doc_tools.cpp [hunk 13]
```diff
@@ -76,241 +76,275 @@ static String _translate_doc_string(const String &p_text) {
 	return translated.indent(indent);
 }
 
-void DocTools::merge_from(const DocTools &p_data) {
-	for (KeyValue<String, DocData::ClassDoc> &E : class_list) {
-		DocData::ClassDoc &c = E.value;
+// Comparator for constructors.
+struct ConstructorCmp {
```
Comment (nlupugla) [inline editor/doc_tools.cpp:80]: I'd rather you we spelled out `Comparator` instead of `Cmp`. Doing so also means we can get rid of the above comment.

File: editor/doc_tools.cpp [hunk 14]
```diff
@@ -76,241 +76,275 @@ static String _translate_doc_string(const String &p_text) {
...
-							not_the_same = true;
+			{
+				// Since constructors can repeat, we need to check the type of
+				// the arguments so we make sure they are different.
+				int64_t arg_count = from.arguments.size();
+				Vector<bool> arg_used;
+				arg_used.resize_zeroed(arg_count);
+				// Also there is no guarantee that argument ordering will match,
+				// so we have to check one by one so we make sure we have an exact match.
+				for (int64_t arg_i = 0; arg_i < arg_count; ++arg_i) {
```
Comment (nlupugla) [inline editor/doc_tools.cpp:147]: is `++arg_i` as opposed to `arg_i++` important here? If not, I'd think you'd want to use `arg_i++` just to be consistent with the rest of the code.

File: editor/doc_tools.cpp [hunk 15]
```diff
@@ -76,241 +76,275 @@ static String _translate_doc_string(const String &p_text) {
...
-					if (not_the_same) {
-						continue;
+				}
+				bool not_the_same = false;
+				for (int64_t arg_i = 0; arg_i < arg_count; ++arg_i) {
+					if (!arg_used[arg_i]) { // At least one of the arguments was different.
+						not_the_same = true;
+						break;
 					}
 				}
```
Comment (nlupugla) [inline editor/doc_tools.cpp:161]: Could this logic me moved inside the previous `i` loop just after the `j` loop is done?

File: editor/doc_tools.cpp (hunk 14 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:147]: This is just the original code

File: editor/doc_tools.cpp (hunk 15 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:161]: I'm leaving the code as it was for now as it's complicated and spaghetti

File: editor/doc_tools.cpp (hunk 13 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:80]: Agreed, got it backwards and thought this was the established wording

File: editor/doc_tools.cpp (hunk 14 above)
Comment (nlupugla) [reply editor/doc_tools.cpp:147]: Maybe I'm misreading the git diff, but I'm seeing lots of red lines that have e.g. `i++` and `j++` in them, so that's what I what I was referring to for being consistent.

File: editor/doc_tools.cpp [hunk 16]
```diff
@@ -76,241 +76,275 @@ static String _translate_doc_string(const String &p_text) {
...
+static void merge_methods(Vector<DocData::MethodDoc> &p_to, const Vector<DocData::MethodDoc> &p_from) {
+	// Get data from `p_to`, to avoid mutation checks. Searching will be done in the sorted `p_to` from the (potentially) unsorted `p_from`.
+	DocData::MethodDoc *to_ptrw = p_to.ptrw();
+	int64_t to_size = p_to.size();
 
-			for (int j = 0; j < cf.methods.size(); j++) {
-				if (cf.methods[j].name != m.name) {
-					continue;
-				}
+	SearchArray<DocData::MethodDoc, MethodCmp> search_array;
```
Comment (nlupugla) [inline editor/doc_tools.cpp]: Ah, I finally see why the `*Cmp` structs need to be structs and not just a static function. Maybe write something like
`typedef MethodSearcher SearchArray<DocData::MethodDoc, MethodCmp>` just after the definition of `MethodCmp` so it's immediately clear why it's being done this way?

File: editor/doc_tools.cpp [hunk 17]
```diff
@@ -76,241 +76,275 @@ static String _translate_doc_string(const String &p_text) {
...
-				m.description = mf.description;
-				m.is_deprecated = mf.is_deprecated;
-				m.deprecated_message = mf.deprecated_message;
-				m.is_experimental = mf.is_experimental;
-				m.experimental_message = mf.experimental_message;
-				m.keywords = mf.keywords;
-				break;
-			}
+		// Check found entry on name.
+		if (to.name == from.name) {
```
Comment (nlupugla) [inline editor/doc_tools.cpp:194]: Is this check redundant with the searching above? Wouldn't the searcher only find a match if their names were equal?

**Review by nlupugla — COMMENTED** [review]
Looks sweet overall! Some great quality of life improvements for this code I'd say :)

Does Godot have a type to represent sorted arrays? It might be nice so it's easier to keep track of assumptions about whether or not an array has already been sorted or not.

If it doesn't, would it be worth putting in some `assert` statements to ensure the array is sorted when it is expected to be, at least in debug builds?

File: editor/doc_tools.cpp (hunk 17 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:194]: It's a binary search, so not guaranteed, it just ensures `from[found]` is *no greater* than `to`

File: editor/doc_tools.cpp (hunk 14 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:147]: But that's consistent with code that doesn't exist any longer :) The code that *is* in the result is consistent

With simple types it doesn't matter if you do pre- or postincrement, but for the iterator it's not necessarily trivial, which is why I used it originally, but it doesn't matter here, and it's far away from any other code to be consistent with, there's no code specification for which is preferred (but preincrement is never going to be worse performance wise than postincrement, when it does matter)

File: editor/doc_tools.cpp [hunk 18]
```diff
@@ -76,241 +76,275 @@ static String _translate_doc_string(const String &p_text) {
...
-						continue;
-					}
-				}
+static void merge_operators(Vector<DocData::MethodDoc> &p_to, const Vector<DocData::MethodDoc> &p_from) {
+	// Get data from `p_to`, to avoid mutation checks. Searching will be done in the sorted `p_to` from the (potentially) unsorted `p_from`.
+	DocData::MethodDoc *to_ptrw = p_to.ptrw();
+	int64_t to_size = p_to.size();
 
-				const DocData::MethodDoc &mf = cf.operators[j];
+	SearchArray<DocData::MethodDoc, OperatorCmp> search_array;
```
Comment (AThousandShips) [inline editor/doc_tools.cpp:290]: This somehow failed to be included in the commit, not sure why, will fix and push

File: editor/doc_tools.cpp (hunk 17 above)
Comment (nlupugla) [reply editor/doc_tools.cpp:194]: Ah, we might want to document that somewhere. It's not directly related to this issue, but I looked briefly at the code for `SearchArray` and it wasn't immediately apparent that would be the behavior in the case the query isn't found. I'm sure it's a standard thing for those who know, but for those who don't it's kind of surprising.

**Comment by AThousandShips** [comment]
> Does Godot have a type to represent sorted arrays? It might be nice so it's easier to keep track of assumptions about whether or not an array has already been sorted or not.
>
> If it doesn't, would it be worth putting in some assert statements to ensure the array is sorted when it is expected to be, at least in debug builds?

There isn't a dedicated type for that no, and we'd want to not copy memory around so it wouldn't really be helpful IMO, we are operating on the actual doc data which uses `Vector`, that'd be a bit more specific programming stuff than Godot really uses or needs

**Review by nlupugla — APPROVED** [review]

File: editor/doc_tools.cpp (hunk 17 above)
Comment (AThousandShips) [reply editor/doc_tools.cpp:194]: That's 100% standard binary search though, that's a bog standard algorithm

Also it *is* documented [here](https://docs.godotengine.org/en/stable/classes/class_packedbytearray.html#class-packedbytearray-method-bsearch), and also [here](https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-method-bsearch)

Comment (nlupugla) [reply editor/doc_tools.cpp:194]: Awesome :D

Comment (AThousandShips) [reply editor/doc_tools.cpp:194]: Granted `SearchArray` is a bit more obscure, but it's very internal and the only reason I use it directly here is because `Vector::bsearch_custom` is non-const and does mutation checks :)

But might be worth documenting that class more directly, will see about that later

Comment (AThousandShips) [reply editor/doc_tools.cpp:194]: Will add notes on those lines to `Vector::bsearch` and the `bisect` method after this :)

**Comment by AThousandShips** [comment]
Realized it doesn't help to sort the loaded data for merge with the general purpose sort algorithm we use, it being sorted or mostly sorted doesn't change anything most likely, haven't studied the algorithm but it's likely a general purpose one with best/worst/average all being `N log N`, would be interesting to look at special purpose sorting algorithms in the future to help with these situations though, but not now

**Comment by AThousandShips** [comment]
Will go over this one a final time this week and squash

**Comment by akien-mga** [comment]
It's been a few weeks since this was made, what's the current status and TL;DR on actual improvements?

IIUC the performance gain may not be that significantly, but is the code quality significantly better? If so, that should be fine.

Are the commits meant to be squashed?

**Comment by AThousandShips** [comment]
I'd say that without a more dynamic sorting algorithm (which we might want to add in the future) the performance improvements are likely marginal, but I think the code quality is greatly improved, will go back over it and re familiarise myself with what this does specifically as it's been quite a while since I touched it

**Comment by AThousandShips** [comment]
The performance remains very much unchanged, it only improves the merge step (about halves the time) but it is so minor that it doesn't matter, but the code is now easier to read in my opinion, future improvements could help with some of these parts I'm sure

**Comment by akien-mga** [comment]
> Should I just add the benchmarks I used to the code? With possible modifications?

Your call, if you expect they'll be useful for further work in the future, they could be added behind a local define.

**Comment by AThousandShips** [comment]
I'll add them if I end up working more on it in the future, for now I think this part isn't something that'll be critical to get good benchmarks on

**Comment by akien-mga** [comment]
Thanks!

**Comment by AThousandShips** [comment]
Thank *you*!
